miniapps/parelag/MultilevelHcurlHdivSolver.cpp

```diff
@@ -0,0 +1,774 @@
+   }
+
+   // Create the solver library.
+   auto lib = SolverLibrary::CreateLibrary(
+                 master_list->Sublist("Preconditioner Library"));
+
+   // Loop through the solvers.
+   for (const auto& solver_name : list_of_solvers)
+   {
+      Timer solver_timer = TimeManager::AddTimer(std::string("Solver \"").
+                                                 append(solver_name).
+                                                 append("\" -- Total"));
+      // Get the solver factory.
+      auto solver_factory = lib->GetSolverFactory(solver_name);
+      auto solver_state = solver_factory->GetDefaultState();
+      solver_state->SetDeRhamSequence(sequence[0]);
+      solver_state->SetBoundaryLabels(ess_attr);
+      solver_state->SetForms({jform});
+
+      // Build the solver.
+      Timer build_timer = TimeManager::AddTimer(std::string("Solver \"").
+                                                append(solver_name).
+                                                append("\" -- Build"));
+      if (!myid)
+      {
+         cout << "Building solver \"" << solver_name << "\"...\n";
+      }
+      unique_ptr<Solver> solver = solver_factory->BuildSolver(A, *solver_state);
+      build_timer.Stop();
+
+      // Run the solver.
+      Timer pre_timer = TimeManager::AddTimer(std::string("Solver \"").
+                                              append(solver_name).
+                                              append("\" -- Pre-solve"));
+      if (!myid)
+      {
+         cout << "Solving system with \"" << solver_name << "\"...\n";
+      }
+
+      // Note that X is on true dofs owned by the process, while x is on local
+      // dofs that are known to the process.
+      Vector X(A->Width()), x(sequence[0]->GetNumberOfDofs(jform));
+      X=0.0;
+
+      {
+         Vector tmp(A->Height());
+         A->Mult(X, tmp);
+         tmp *= -1.0;
+         tmp += B;
+
+         double local_norm = tmp.Norml2();
+         local_norm *= local_norm;
+         double global_norm;
+         MPI_Reduce(&local_norm, &global_norm, 1, GetMPIType(local_norm),
+                    MPI_SUM, 0, comm);
+         if (!myid)
+         {
+            cout << "Initial residual l2 norm: " << sqrt(global_norm) << '\n';
+         }
+      }
+      pre_timer.Stop();
+
+      // Perform the solve.
+      Timer solve_timer = TimeManager::AddTimer(std::string("Solver \"").
+                                                append(solver_name).
+                                                append("\" -- Solve"));
+      solver->Mult(B, X);
+      solve_timer.Stop();
+
+      Timer post_timer = TimeManager::AddTimer(std::string("Solver \"").
+                                               append(solver_name).
+                                               append("\" -- Post-solve"));
+      {
+         Vector tmp(A->Height());
+         A->Mult(X, tmp);
+         tmp *= -1.0;
+         tmp += B;
+
+         double local_norm = tmp.Norml2() * tmp.Norml2();
```

Comment:
Not a big deal, but there are some places with `local_norm *= local_norm` and some places with two calls to `Norml2()`.

Reply:
Ops. I forgot this one. It should be fixed now.